[WIP] Dual Battery Support Enhancements

Sniki opened the pull request:
Thanks @armenio for the SSDT, _BIX method was a requirement for newer laptops.

I added the `If (_OSI ("Darwin"))` code to the `Method _STA` and `Method (_INI)`
(not sure if it is necessary to have it on both methods or just one can do the job),
Corrections appreciated, however i am using it right now as it is and haven't noticed a problem.

By this we ensure that the SSDT doesn't affect other Operating Systems (WIndows, Linux, etc) as this is a macOS specific problem.

Also the description of the binpatch section is wrong as it affects Windows and Linux or any other OS so it would be wise to either:
1. Correct description of SSDT so instead of binpatching BAT0/BAT1 to BATC with a rename, it would be wiser to rename the method(s) containing the notifier(s).
Example of unpatched (Method _Q4A):

```
Method (_Q4A, 0, NotSerialized)
{
    CLPM ()
    Notify (BAT0, 0x81)
}
```
Appropriate way of BAT0 or BAT1 to BATC rename without affecting other OS would be:
Rename Method (_Q4A) to XQ4A and provide the patched method on different SSDT or SSDT-BATC with If (_OSI ("Darwin") and Notify (BATC, 0x81) like this:

```
Method (_Q4A, 0, NotSerialized)
{
    If (_OSI ("Darwin"))
    {
        CLPM ()
        Notify (BATC, 0x81)
    }
    Else
    {
        \_SB.PCI0.LPC.EC.XQ4A ()
    }
}
```
This way we ensure that the Patched method loads under "Darwin" , if not it returns the original method that we renamed as XQ4A in above example.

Correcting and documenting the description of the SSDT-BATC i believe would be pointless and bloat the SSDT-BATC, instead it may be wiser to:

2. Make a separate document on how to implement dual battery support.

This will be a work in progress PR
Comments, critics, help and suggestions are appreciated

Sniki committed:
- Added If (_OSI ("Darwin")) to Method (_INI) and Method (_STA)

Andrey1970AppleLife commented:
CC @armenio

vit9696 commented:
Yes, separating this to a separate document sounds nicer to me. Could you handle that too?

Sniki commented:
> Yes, separating this to a separate document sounds nicer to me. Could you handle that too?

Yes, i am creating a short tutorial right now, will add the commit once i think it's ready for a review.
@vit9696 what do you think about the first commit here, is it necessary in both _STA and _INI or just one of them ?

Sniki committed:
- Added QuickGuide
- Added Samples

Sniki commented:
I finished the 1st draft of QuickGuide and added the samples.
A review would be appreciated, with suggestions included if changes necessary.

Thanks !

Sniki committed:
- Moved files to Docs folder
- Docs: Fix Typos #1

Sniki committed:
Docs: Fix Typos 2

timb3rlandqc commented:
sniki i have done your tuto but when I unplug the ac cable the intensity and the whatts drop too 0 this cause red battery random any idea ?

Sniki commented:
> sniki i have done your tuto but when I unplug the ac cable the intensity and the whatts drop too 0 this cause red battery random any idea ?

Not a troubleshooting thread, please use forums.

timb3rlandqc commented:
thx for the help appreciate

vit9696 commented:
This looks about right to me really, though I am not quite aware of the laptop configuration. In fact I did not even notice the typos fixed in the latest commit. I would like @usr-sse2 to make a brief look on it, but afterwards we can merge. If there are any practical concerns we can of course discuss them.

Sniki commented:
> This looks about right to me really, though I am not quite aware of the laptop configuration. In fact I did not even notice the typos fixed in the latest commit. I would like @usr-sse2 to make a brief look on it, but afterwards we can merge. If there are any practical concerns we can of course discuss them.

Thanks @vit9696 
If you think it would be better, i can also upload the DSDT.aml as sample or maybe even EFI if you need to have a glance on the configuration, just took one of these laptops as sample as they are the most common ones that have dual batteries.
No problem, i am always open for discussions and improvements.

usr-sse2 commented on Dual Battery Support.md:
```diff
@@ -0,0 +1,262 @@
+#  Dual Battery Support
+
+macOS has buggy dual battery support and we won't have correct battery status without doing some changes.
```
Buggy or no support?

Sniki replied:
Buggy IMO,
If you leave the setup unimplemented, in about this mac > Power > it will show both batteries.
However it's totally bugged, no correct reporting, doesn't show both batteries on menu bar etc.
My guess is that long time ago in Yosemite/El Capitan era, they may have had plans for a Macbook or so with Dual batteries and start working on the software side but for any reason cancelled that and they may have not touched that part anymore, at least that's how i think.

usr-sse2 replied:
OK. I see that there are some SMC keys that have battery index, but not all.

usr-sse2 commented on Dual Battery Support.md:
````diff
@@ -0,0 +1,262 @@
+Notify (\_SB.PCI0.LPCB.EC.BAT1, 0x03)
+```
+
+In order for our SSDT-BATC to work, we should rename these notifiers to BATC, in our example:
````
Renaming is when you change the name of a device (e. g. HECI to IMEI) and all references to it. Here we're changing references from individual batteries to a combined one, so we should say *change* instead of *rename*.

Sniki replied:
Thanks for the hint, ok will do so in next upcoming commit.

Sniki replied:
Resolved in latest commit

usr-sse2 commented on Dual Battery Support.md:
```diff
@@ -0,0 +1,262 @@
+This makes sure that the patched code does get executed only if the operating system is macOS ("Darwin").
+So now the code will execute on macOS but won't in other Operating Systems like Windows or Linux, etc.
+For this we add the Else into the code and call the original method that we renamed to Method XQ22.
+By this, if macOS, the patched SSDT will load, if it's something else like Windows or Linux it will load the original method that we renamed to XQ22 in our case, like shown below:
```
All SSDTs will *load* in all operating systems. The original method will be *executed* on Windows and Linux, as the `If (_OSI ("Darwin"))` is a runtime condition.

Sniki replied:
Ok will cleanup this section as well

Sniki replied:
@usr-sse2 edited this section on latest commit, let me know if it needs more tweaking.

Sniki committed:
Docs: Fixes as requested by @usr-sse2

armenio commented:
Since there is already the implementation of SSDT-BATC.dsl is ready,
There is no need for it to be copied to BATC-Sample.dsl (this even can cause confusion)
As a common programming practice, is a good idea to keep separate files, avoiding code duplication and maintenance, just as it is done with the basic battery patch (breaking fields into 8-bit parts).
It is necessary to consider that at some point SSDT-BATC.dsl can also be updated more easily
Of course, afterwards this can be added to a file in just one "automated build" as a personal way to use the hack.

Sniki commented:
> Since there is already the implementation of SSDT-BATC.dsl is ready,
> There is no need for it to be copied to BATC-Sample.dsl (this even can cause confusion)
> As a common programming practice, is a good idea to keep separate files, avoiding code duplication and maintenance, just as it is done with the basic battery patch (breaking fields into 8-bit parts).
> It is necessary to consider that at some point SSDT-BATC.dsl can also be updated more easily
> Of course, afterwards this can be added to a file in just one "automated build" as a personal way to use the hack.

The sample includes the Patched Methods that notify BATC in the end and the purpose and usage is mentioned on the Dual Battery Support.md for each.

SSDT-BATC is the actual SSDT that we will need, BATC-Sample is the one with the patched methods included on it.
So basically aside from the 'If (_OSI ("Darwin"))' that is added to the SSDT-BATC, the rest is clean as it was previously.
But if you think it would be better, i can cleanup BATC-Sample.dsl so it does have only the patched notifiers for BATC.

armenio commented:
> > Since there is already the implementation of SSDT-BATC.dsl is ready,
> > There is no need for it to be copied to BATC-Sample.dsl (this even can cause confusion)
> > As a common programming practice, is a good idea to keep separate files, avoiding code duplication and maintenance, just as it is done with the basic battery patch (breaking fields into 8-bit parts).
> > It is necessary to consider that at some point SSDT-BATC.dsl can also be updated more easily
> > Of course, afterwards this can be added to a file in just one "automated build" as a personal way to use the hack.
> 
> The sample includes the Patched Methods that notify BATC in the end and the purpose and usage is mentioned on the Dual Battery Support.md for each.
> 
> SSDT-BATC is the actual SSDT that we will need, BATC-Sample is the one with the patched methods included on it.
> So basically aside from the 'If (_OSI ("Darwin"))' that is added to the SSDT-BATC, the rest is clean as it was previously.
> But if you think it would be better, i can cleanup BATC-Sample.dsl so it does have only the patched notifiers for BATC.

From a programmer's point of view, I don't really see the need for code duplication. this is not good practice.

I understood your example perfectly, but (in my personal opinion) it is better to suggest this: "In addition to using SSDT-BATC.dsl, you also need ...", this can be in the Sample.dsl header as well. because in that case, it makes maintenance viable both for SSDT-BATC.dsl and also for specific patches.

Andrey1970AppleLife commented:
Delete BATC-Sample.dsl , I agree with @armenio

Sniki commented:
> Delete BATC-Sample.dsl , I agree with @armenio

Will do tonight, i will also need to cleanup the md file from its references to that file.

Sniki committed:
Docs: Fixes as requested

- Cleaned up and adjusted the SSDT-BATC comments
- Added sample SSDT with only patched notifiers into the Dual Battery Support.md
- Removed BATC-Sample.dsl

Sniki commented:
Done the changes as requested.

- Removed the BATC-Sample.dsl and added just the patched notifiers as a sample into the md file
- Adjusted the documentation and fixed some typos.

Let me know if anything else is needed.
Might consider a similar **"How to patch ACPI for battery status support"** in the future.

vit9696 commented:
Merged, tyvm. Other guides will always be appreciated as well I guess =)

Juan-VC-gh commented:
I have wanted to implement both batteries for my laptop for a long time and I can now try this guide! I got a question my laptop isn't that old but devices BAT0 and BAT1 do not have the _BIX method, should I remove it from SSDT-BATC?

Edit: I tried without removing BIX because comments point that it adds the support for VirtualSMC but had no luck, I still see only a battery with the sensei app that's why it shows only 68% health. Attached all battery-related files that need to be modified + dsdt.dsl if anyone minds looking at it.
[DBAT.zip](https://github.com/acidanthera/VirtualSMC/files/4828338/DBAT.zip)

Andrey1970AppleLife commented:
@Sniki, move BATC-Sample.plist and Dual Battery Support.md to Docs folder.

vit9696 commented:
I will better move that myself later along with the 11.0 batch commit.
